version_test.go

```diff
@@ -32,6 +32,7 @@ func TestNewVersion(t *testing.T) {
 		{"v1.2.3-rc1-with-hypen", false},
 		{"1.2.3.4", true},
 		{"v1.2.3.4", true},
+		{"20161202202307-sha.e8fc5e5", false},
```

Comment:
The test you're using isn't really semver. What about using the example `v2.3.5-20161202202307-sha.e8fc5e5` from the ticket?

On the tip of master the test for `v2.3.5-20161202202307-sha.e8fc5e5` passes for this test. There should be a constraints test for this which fails for this PR.

It's the constraint handling rather than the version handling that's failing for this case.